include/ginkgo/core/base/abstract_factory.hpp

```diff
@@ -435,14 +443,14 @@ class deferred_factory_parameter {
  * @ingroup LinOp
  */
 #define GKO_FACTORY_PARAMETER(_name, ...)                                    \
-    mutable _name{__VA_ARGS__};                                              \
+    _name{__VA_ARGS__};                                                      \
                                                                              \
     template <typename... Args>                                              \
-    auto with_##_name(Args&&... _value)->std::decay_t<decltype(*this)>&      \
+    parameters_type& with_##_name(Args&&... _value)                          \
```

Comment:
But shouldn't we rather fix the deferred_factory macro then? I think the old way should be usable as mixin without the type alias.

Reply:
If you want to keep the new approach, then you must also modify the `GKO_CREATE_FACTORY_PARAMETERS` and `GKO_ENABLE_LIN_OP_FACTORY` macro to fix the name of the parameters type to `parameters_type`.

Reply:
Ok, I understand the issue, my mind was at a different place. I think then changing the macros would be the best approach. But I think the whole change might be interface breaking. I would be fine with that, since, as you mentioned, the current implementation is not correct for the mixins anyway.

Reply:
I check some base type of `parameter_types`, self is introduced in `enable_parameter_type` `protected` scope already.
All parameter types are inherited from that, so we always have `self()` actually.
I just need to delete the `GKO_ENABLE_SELF()` in the two parameters collection in solver_bases.hpp and then I can use `self()`
I am trying to see this approach based on `self()` gives any trouble